FLT/AutomorphicForm/QuaternionAlgebra/HeckeOperators.lean

```diff
@@ -0,0 +1,182 @@
+import Mathlib.Algebra.BigOperators.GroupWithZero.Action
+import Mathlib.GroupTheory.GroupAction.Quotient
+import Mathlib
+/-
+
+# Abstract Hecke operators
+
+We give an abstract exposition of the theory of Hecke operators
+
+The set-up: a group `G` acts on additive group `A`, we have
+an element `g : G`, and `U`, `V` are subgroups of `G`. We impose the
+finiteness hypothesis that the double coset `UgV` is a *finite* union
+of single left cosets `gᵢV`. Under this hypothesis we define a Hecke
+operator [UgV] or `T_g`, which is an additive group homorphism
+from `A^V` (the `V`-fixedpoints of `G` on `A`) to `A^U`.
+
+## Main definition
+
+## Mathematical details
+
+
+Before we give the definition, let us observe that our finiteness
+hypothesis is the same as asking that g⁻¹Ug ∩ V has finite index in
+g⁻¹Ug, and by an appropriate version of the second isomorphism theorem,
+that g⁻¹UgV is a finite union of left cosets hᵢV of V.
+Hence the double coset UgV is also a finite union of left cosets
+of `V` (namely `ghᵢV`).
+
+The definition of the Hecke operator is as follows. Write UgV as a
+finite disjoint union gᵢV.
+If a ∈ A^V then we define `[UgV]a := ∑ᵢ gᵢ•a`. Note that replacing
+the choice of gᵢ with another element g'ᵢ := gᵢv will not change gᵢ•a
+as a ∈ A^v, so the sum is a well-defined element of A. Finally
+we observe that it's in A^U because if u ∈ U then left multiplication
+by u is a permutation of the cosets gᵢV.
+
+Note that if G is a topological group and U, V are compact open
+subgroups of G, then our hypothesis is automatically satisfied
+for all g ∈ G, because g⁻¹Ug ∩ V is open in compact V and hence
+has finite index.
+
+-/
+
+section move_these
+
+-- This should be in Mathlib.Algebra.BigOperators.Finprod
+lemma finsum_mem_eq_finset_sum {ι A : Type*} [AddCommMonoid A]
+    {s : Set ι} (hs : s.Finite) (f : ι → A) :
+    ∑ᶠ i ∈ s, f i = ∑ i ∈ hs.toFinset, f i := by
+  rw [finsum_mem_eq_sum_of_inter_support_eq]
+  simp
```

Comment:
This is `finsum_mem_eq_finite_toFinset_sum` (quite the mouthful), and it's in `Mathlib.Algebra.BigOperators.Finprod`

Reply:
Thanks; removed.